smart_open/ftp.py

```diff
@@ -0,0 +1,107 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2019 Radim Rehurek <me@radimrehurek.com>
+#
+# This code is distributed under the terms and conditions
+# from the MIT License (MIT).
+#
+
+"""Implements I/O streams over FTP.
+"""
+
+import getpass
+import logging
+import urllib.parse
+import smart_open.utils
+from ftplib import FTP
+
+logger = logging.getLogger(__name__)
+
+SCHEME = "ftp"
+
+"""Supported URL schemes."""
+
+DEFAULT_PORT = 21
+
+URI_EXAMPLES = (
+    "ftp://username@host/path/file",
+    "ftp://username:password@host/path/file",
+    "ftp://username:password@host:port/path/file",
+)
+
+
+def _unquote(text):
+    return text and urllib.parse.unquote(text)
+
+
+def parse_uri(uri_as_string):
+    split_uri = urllib.parse.urlsplit(uri_as_string)
+    assert split_uri.scheme in SCHEME
+    return dict(
+        scheme=split_uri.scheme,
+        uri_path=_unquote(split_uri.path),
+        user=_unquote(split_uri.username),
+        host=split_uri.hostname,
+        port=int(split_uri.port or DEFAULT_PORT),
+        password=_unquote(split_uri.password),
+    )
+
+
+def open_uri(uri, mode, transport_params):
+    smart_open.utils.check_kwargs(open, transport_params)
+    parsed_uri = parse_uri(uri)
+    uri_path = parsed_uri.pop("uri_path")
+    parsed_uri.pop("scheme")
+    return open(uri_path, mode, transport_params=transport_params, **parsed_uri)
+
+
+def convert_transport_params_to_args(transport_params):
+    supported_keywords = [
+        "timeout",
+        "source_address",
+        "encoding",
+    ]
+    unsupported_keywords = [k for k in transport_params if k not in supported_keywords]
+    kwargs = {k: v for (k, v) in transport_params.items() if k in supported_keywords}
+
+    if unsupported_keywords:
+        logger.warning(
+            "ignoring unsupported ftp keyword arguments: %r", unsupported_keywords
+        )
+
+    return kwargs
+
+
+def _connect(hostname, username, port, password, transport_params):
+    kwargs = convert_transport_params_to_args(transport_params)
+    ftp = FTP(**kwargs)
+    try:
+        ftp.connect(hostname, port)
+        ftp.sendcmd(f"USER {username}")
+        ftp.sendcmd(f"PASS {password}")
+    except Exception:
+        raise Exception(
+            "Unable to login to FTP server: either username, password, host, or port is wrong!"
+        )
+    return ftp
+
+
+# transport paramaters can include any extra parameters that you want to be passed into FTP_TLS
+def open(
+    path,
+    mode="r",
+    host=None,
+    user=None,
+    password=None,
+    port=DEFAULT_PORT,
+    transport_params=None,
+):
+    if not host:
+        raise ValueError("you must specify the host to connect to")
+    if not user:
+        user = getpass.getuser()
+    if not transport_params:
+        transport_params = {}
+    conn = _connect(host, user, port, password, transport_params)
+    fobj = conn.transfercmd(f"RETR {path}").makefile(mode)
```

Comment:
`RETR` is only used for reading a file. I think we should add support for writing to a file on a FTP server as well

Reply:
That is a good question. The entire file is not downloaded into memory. In fact, when open is called, it only returns a file like object representing the file on the FTP server, but it will not have downloaded any part of that file into memory. How the file is downloaded depends on what the developer wants to do with the object returned by `open`. For example:

```
# At this point, the file has not been downloaded at all
ftp_file_obj = open("ftp://username:password@host:port/file)

# Now, the file object could be read line by line (this is where the file from the ftp server is actually read)
for line in ftp_file_obj:
    print(line)
```

More specifically: `conn.transfercmd(f"RETR {path}")` returns a [socket object](https://docs.python.org/3/library/socket.html#socket.socket.makefile). When created, this object doesn't download any part of the file. You can see an example of the makefile function being called on a socket object [here](https://github.com/python/cpython/blob/main/Lib/ftplib.py#L463).

Reply:
Good find and good to know we are streaming the file. In the ftplib.py file you linked it looks like `retrlines` has quite a lot of useful functionality for reading a file that we don't get with our current implementation though such as ensuring that we close the socket you mentioned so we don't leak resources and ensuring that we handle various newlines and CLRF's properly. To me it seems like we should use the higher level `retrlines` method with all of the functionality baked in